test/e2e/cluster_deletion.go

```diff
@@ -0,0 +1,399 @@
+				Union(phaseInput.DeleteBlockKinds).
+				Union(phaseInput.DeleteKinds)
+		}
+
+		// Get all objects relevant to the test by filtering for the kinds of the phases.
+		graph, err := clusterctlcluster.GetOwnerGraph(ctx, namespace.GetName(), input.BootstrapClusterProxy.GetKubeconfigPath(), func(u unstructured.Unstructured) bool {
+			return relevantKinds.Has(u.GetKind())
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+		for i, phase := range input.ClusterDeletionPhases {
+			// Iterate over the graph and find all objects relevant for the phase.
+			// * Objects matching the kind except for Machines (see below)
+			// * Machines which are owned by one of the kinds which are part of the phase
+			allKinds := phase.DeleteBlockKinds.Union(phase.DeleteKinds)
+			allKindsWithoutMachine := allKinds.Clone().Delete("Machine")
+
+			for _, node := range graph {
+				// Only add objects which match a defined kind for the phase.
+				if !allKinds.Has(node.Object.Kind) {
+					continue
+				}
+
+				// Special handling for machines: only add machines which are owned by one of the given kinds.
+				if node.Object.Kind == "Machine" {
+					isOwnedByKind := false
+					for _, owner := range node.Owners {
+						if allKindsWithoutMachine.Has(owner.Kind) {
+							isOwnedByKind = true
+							break
+						}
+					}
+					// Ignore Machines which are not owned by one of the given kinds.
+					if !isOwnedByKind {
+						continue
+					}
+				}
+
+				obj := new(unstructured.Unstructured)
+				obj.SetAPIVersion(node.Object.APIVersion)
+				obj.SetKind(node.Object.Kind)
+				obj.SetName(node.Object.Name)
+				obj.SetNamespace(namespace.GetName())
+
+				phase.objects = append(phase.objects, obj)
+
+				// If the object's kind is part of kindsWithFinalizer, then additionally the object to the finalizedObjs slice.
+				if phase.DeleteBlockKinds.Has(obj.GetKind()) {
+					objectsWithFinalizer = append(objectsWithFinalizer, obj)
+				}
+			}
+
+			// Update the phase in input.ClusterDeletionPhases
+			input.ClusterDeletionPhases[i] = phase
+		}
+
+		// Update all objects in objectsWithFinalizers and add the finalizer for this test to control when these objects vanish.
+		By(fmt.Sprintf("Adding finalizer %s on objects to block during cluster deletion", finalizer))
+		addFinalizer(ctx, input.BootstrapClusterProxy.GetClient(), finalizer, objectsWithFinalizer...)
+
+		// Trigger the deletion of the Cluster.
+		framework.DeleteCluster(ctx, framework.DeleteClusterInput{
+			Cluster: clusterResources.Cluster,
+			Deleter: input.BootstrapClusterProxy.GetClient(),
+		})
+
+		var objectsDeleted, objectsInDeletion []client.Object
+
+		for i, phase := range input.ClusterDeletionPhases {
+			By(fmt.Sprintf("Verify deletion phase %d/%d", i+1, len(input.ClusterDeletionPhases)))
+			// Objects which previously were in deletion because they waited for our finalizer being removed are now checked to actually had been removed.
+			objectsDeleted = objectsInDeletion
```

Comment:
Maybe it would be slightly better if we add an "expected" prefix to these 3 variables